### PR TITLE
Report multiple build causes for image change triggers

### PR DESCRIPTION
**Pull request by smarterclayton**
Improve tests

[test]

@bparees this is not a fix to #14725, but increases coverage and is more accurate.

### CODE REVIEW

File: pkg/image/trigger/buildconfigs/buildconfigs.go [hunk 1]
```diff
@@ -141,43 +138,64 @@ func (r *BuildConfigReactor) ImageChanged(obj interface{}, tagRetriever trigger.
...
 			continue
 		}
-		if latest != p.LastTriggeredImageID {
-			changed = true
-			ref = latest
-			break
+
+		// prevent duplicate build trigger causes
+		if fired == nil {
+			fired = make(map[kapi.ObjectReference]string)
```
Comment (bparees) [inline pkg/image/trigger/buildconfigs/buildconfigs.go]: why not initialize this where you declared the variable? going for efficiency?

**Comment by smarterclayton** [comment]
If 3 images change simultaneously, and the build has triggers for all 3, then we will go through 3 distinct buildrequests, each one containing one of the images as the "triggeredbyimage" value.

Um... why would we do that?

**Comment by smarterclayton** [comment]
We currently resolve all triggers at build instantiate time - so we would not trigger 3.

**Comment by bparees** [comment]
> We currently resolve all triggers at build instantiate time - so we would not trigger 3.

ok yeah, i forgot we do that, however it's still the case that what is happening is we are running a build that explicitly picks up a particular image, and then implicitly picks up the rest.

Meaning that the build is guaranteed to use the one image listed in the triggeredbyimage section (and what was supposed to be listed in the triggeredby section), it is not guaranteed to use all N images that you are now listing in the triggeredby section.  Essentially you're just using the triggeredby section as "other stuff that was also different when we ran this build.  the build may pick these up, if they are not replaced with something even newer before this build runs".

If you prefer that, ok, it's certainly potentially useful information, but the implication/meaning is definitely a bit fuzzy.

**Comment by bparees** [comment]
(well, "before this build runs" is an overstatement since the resolution happens during instantiate, not when the build runs.  so there's almost no chance the build runs with anything different than what you're going to report, but it is non-zero)

**Comment by smarterclayton** [comment]
I think from an end user perspective, id want to see all of the causes listed.

**Comment by bparees** [comment]
> I think from an end user perspective, id want to see all of the causes listed.

then i guess my only open question is the way you're initializing the map, but i assume you're doing that to save initializing it when we don't enter the condition.

**Comment by smarterclayton** [comment]
Right, and to avoid doing a second resolution on an image and getting
a different one.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Any other comments?

**Comment by bparees** [comment]
lgtm

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[severity:bug]

**Comment by deads2k** [comment]
re[test]

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Pull request merged**